### PR TITLE
[Move IR] Make explicit deps lazy

### PR DESCRIPTION
**Pull request by tnowacki**
- Explicit deps in the IR were loaded into the pools eagerly
- They are now added to the existing compiled module dep list and then used lazily (like the other deps)

## Motivation

- The old way of eager loading the explicit dependency list into the pools was brittle. Any change in the ordering of those members caused a change in the compiled module
- This was an issue as the compiler toposorts the dependencies. In on case, changing something in a transitive dependency (a grandparent module of sorts) would result in the module being changed because the dependencies are now in a different order (a still correct but different ordering)
- Switching to a lazy loading means the tables will only change when the modules code changes, which is what one would expect. 

## Test Plan

- ran tests
- There used to be some test around the stdlib compiled modules... not sure if that is still there

### CODE REVIEW

File: language/compiler/ir-to-bytecode/src/compiler.rs [hunk 1]
```diff
@@ -563,15 +579,27 @@ pub fn compile_module<'a, T: 'a + ModuleAccess>(
...
+    let mut compiled_deps = vec![];
     for dependency in dependencies {
         let ModuleDependency {
             name: mname,
             structs,
             functions,
         } = dependency;
+        let current_module = outer_context.module_ident(&mname)?;
+        let mut context = Context::new(outer_depedencies.clone(), Some(current_module.clone()))?;
+        // TODO find a way to accumulate depedencies. Borrow issues galore on attempts
```
Comment (bob-wilson) [inline language/compiler/ir-to-bytecode/src/compiler.rs:597]: Did you intend to fix this before landing?

Comment (tnowacki) [reply language/compiler/ir-to-bytecode/src/compiler.rs:597]: Not really :P 

I think I might have an idea (gained from working on another PR) but it is a really tricky thing to fix

Comment (tnowacki) [reply language/compiler/ir-to-bytecode/src/compiler.rs]: @zekun000 showed me the rental crate which let me solve this issue. I added another commit to this PR with the fix

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A rental 0.5.5 (direct third-party, crates.io)
    * features: default, std
  A stable_deref_trait 1.2.0 (transitive third-party, crates.io)
    * features: alloc, std

host packages:
  A rental-impl 0.5.5 (transitive third-party, crates.io)
    * features: [none]
  M syn 1.0.58 (direct third-party, crates.io)
    * added features: fold
    * (unchanged features: clone-impls, default, derive, extra-traits, full, parsing, printing, proc-macro, quote, visit, visit-mut)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A rental 0.5.5 (direct third-party, crates.io)
    * features: default, std
  A stable_deref_trait 1.2.0 (transitive third-party, crates.io)
    * features: alloc, std

host packages:
  A rental-impl 0.5.5 (transitive third-party, crates.io)
    * features: [none]
  M syn 1.0.58 (direct third-party, crates.io)
    * added features: fold
    * (unchanged features: clone-impls, default, derive, extra-traits, full, parsing, printing, proc-macro, quote, visit, visit-mut)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A rental 0.5.5 (direct third-party, crates.io)
    * features: default, std
  A stable_deref_trait 1.2.0 (transitive third-party, crates.io)
    * features: alloc, std

host packages:
  A rental-impl 0.5.5 (transitive third-party, crates.io)
    * features: [none]
  M syn 1.0.60 (direct third-party, crates.io)
    * added features: fold
    * (unchanged features: clone-impls, default, derive, extra-traits, full, parsing, printing, proc-macro, quote, visit, visit-mut)
```

File: language/compiler/ir-to-bytecode/src/compiler.rs [hunk 2]
```diff
@@ -562,16 +574,26 @@ pub fn compile_module<'a, T: 'a + ModuleAccess>(
...
+// and sets them after a successful result
+// Any `Error` should stop compilation in the caller
 fn compile_explicit_dependency_declarations(
-    context: &mut Context,
+    outer_context: &mut Context,
+    address_opt: Option<AccountAddress>,
+    imports: Vec<ImportDefinition>,
     dependencies: Vec<ModuleDependency>,
 ) -> Result<()> {
+    let mut depedencies_acc = outer_context.take_dependencies();
```
Comment (bob-wilson) [inline language/compiler/ir-to-bytecode/src/compiler.rs:586]: spelling? "depedencies"

File: language/compiler/ir-to-bytecode/src/context.rs [hunk 3]
```diff
@@ -154,6 +158,51 @@ impl<'a> CompiledDependency<'a> {
...
+    pub fn create(module: CompiledModule) -> Result<Self> {
+        Self::try_new(Box::new(module), move |module| {
+            CompiledDependencyView::new(module)
+        })
+        .map_err(|e| e.0)
+    }
+}
+
+pub(crate) enum CompiledDependency<'a> {
+    /// Simple `CompiledDependecyView` where the borrowed `CompiledModule` is held elswehere,
```
Comment (bob-wilson) [inline language/compiler/ir-to-bytecode/src/context.rs:186]: "elswehere"

File: language/compiler/ir-to-bytecode/src/context.rs [hunk 4]
```diff
@@ -154,6 +158,51 @@ impl<'a> CompiledDependency<'a> {
...
+    /// Normally it is borrowed from outside of the compilers API
+    Borrowed(CompiledDependencyView<'a>),
+    /// `Stored` holds the `CompiledModule` as well as the `CompiledDependencyView` into the module
+    /// uses `rental` for a self referntial struct
```
Comment (bob-wilson) [inline language/compiler/ir-to-bytecode/src/context.rs:190]: "referntial"

**Review by bob-wilson — APPROVED** [review]
I'm glad you found a way to fix the TODO. The solution with the rental crate isn't entirely clear to me, so some extra comments around that would be nice.

Oh, and I found some comment typos 🤪

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A rental 0.5.5 (direct third-party, crates.io)
    * features: default, std
  A stable_deref_trait 1.2.0 (transitive third-party, crates.io)
    * features: alloc, std

host packages:
  A rental-impl 0.5.5 (transitive third-party, crates.io)
    * features: [none]
  M syn 1.0.60 (direct third-party, crates.io)
    * added features: fold
    * (unchanged features: clone-impls, default, derive, extra-traits, full, parsing, printing, proc-macro, quote, visit, visit-mut)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A rental 0.5.5 (direct third-party, crates.io)
    * features: default, std
  A stable_deref_trait 1.2.0 (transitive third-party, crates.io)
    * features: alloc, std

host packages:
  A rental-impl 0.5.5 (transitive third-party, crates.io)
    * features: [none]
  M syn 1.0.60 (direct third-party, crates.io)
    * added features: fold
    * (unchanged features: clone-impls, default, derive, extra-traits, full, parsing, printing, proc-macro, quote, visit, visit-mut)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A rental 0.5.5 (direct third-party, crates.io)
    * features: default, std
  A stable_deref_trait 1.2.0 (transitive third-party, crates.io)
    * features: alloc, std

host packages:
  A rental-impl 0.5.5 (transitive third-party, crates.io)
    * features: [none]
  M syn 1.0.60 (direct third-party, crates.io)
    * added features: fold
    * (unchanged features: clone-impls, default, derive, extra-traits, full, parsing, printing, proc-macro, quote, visit, visit-mut)
```

**Pull request closed without merging**